Add "All Permissions" expression

supertassu opened the pull request:
Target Minecraft versions: **any**
Requirements: **none**
Related issues: #900

Description:
This PR adds a way to loop all permission that a player has,

Test Skript:

```
on join:
    set {_perms::*} to all permissions of player
    message "Welcome %player%! You have %amount of {_perms::*}% permissions."
    message "Here they are: "
    loop player's permission:
        message " -> %loop-value%"

    # didn't break anything
    if player has permission "bukkit.command.me": # a default permission
        message "You have permission ""bukkit.command.me"""
    else:
        message "You do not have permission ""bukkit.command.me"""

```

ps. it worked just fine in IntellJ IDEA.

Blueyescat commented:
Looks like so useless, it needs support for all changers. Also the syntaxes should be:
```vb
[all] [the] permissions of %players%
%players%'[s] [all] permissions
```

supertassu commented:
> Looks like so useless, it needs support for all changers.

What do you mean?

> Also the syntaxes should be

Thanks, fixing.

Blueyescat commented:
Changers = set, add, remove, reset, clear, delete etc.
So you can edit permissions, what can you do with a list of permissions :stuck_out_tongue:

Also %players% in syntax means you can do ``permissions of all players``, did you do something about that?

bensku commented:
Editing permissions is quite hard, and might break other permission plugins. It should be left out, unless we really test with all popular ones and verify it works.

supertassu commented:
Permissions of all players is good catch. 

> what can you do with a list of permissions 😛

Issue #900 has an usage case.

Blueyescat commented:
Okay, but add multi players support at least, Skript does this every time. Also what is allowing to "player permission" with these syntaxes... It is **``'[s]``** everywhere.

supertassu commented:
@Blueyescat I've added support for multiple players in the expression. Here's another test skript:

```
on chat: # just an event
        loop all players:
                add loop-player to {_allplayers::*}
                add loop-player to {_allplayers::*} # too lazy to get another minecraft instance running
                broadcast "debug: added %loop-player% to the list twice. List currently has %size of {_allplayers::*}% entries"
        broadcast all permissions of {_allplayers::*}
```

TheBentoBox commented:
I want to note that these syntaxes aren't the best:
```
[all] [the] permission[s] of %players%
%players%'[s] [all] permission[s]
```
The first one allows for incorrect English (`all the permissions of player` is not correct), and the `all` in the second one is also blatantly incorrect English (`player's all permissions` doesn't make sense). They should probably just be something like this:
```
[(all [of]|[the])] permissions of %players%
%players%'[s] permissions
```

supertassu commented:
@TheBentoBox Fixed (again).

Blueyescat commented:
hmm, then this syntax is incorrect too :/
```vb
[all] [the] [book] (pages|content) of %item stack%
%item stack%'s [book] (pages|content)
[book] page %number% of %item stack%
%item stack%'s [book] page %number%```

TheLimeGlass commented:
It should be like
```
@Name("Permissions")
@Description("Grabs a string list of all permissions a player has.")
@Examples({"set {_permissions::*} to permissions of player"})
@Since("2.2-dev33")
public class Permissions extends PropertyExpression<Player, String> {
	
	static {
		Skript.registerExpression(Permissions.class, String.class, ExpressionType.SIMPLE, "[the] permissions of %players%", "%players%['s] permissions");
	}

	@SuppressWarnings("unchecked")
	@Override
	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
		setExpr((Expression<? extends Player>) exprs[0]);
		return true;
	}
	
	@Override
	protected String[] get(final Event event, final Player[] source) {
		final ArrayList<String> permissions = new ArrayList<>();
		for (final Player player : getExpr().getArray(event)) 
			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
				permissions.add(permission.getPermission());
		return permissions.toArray(new String[permissions.size()]);
	}
	
	@Override
	@Nullable
	public Class<?>[] acceptChange(final ChangeMode mode) {
		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.SET) 
			return super.acceptChange(mode);
		return null;
	}
	
	@Override
	public void change(final Event event, final @Nullable Object[] delta, final ChangeMode mode) {
		assert delta != null;
		for (final Player player : getExpr().getArray(event))
			for (final PermissionAttachmentInfo attachment : player.getEffectivePermissions())
				for (final String permission : (String[]) delta)
					attachment.getAttachment().setPermission(permission, mode == ChangeMode.ADD);
	}
	
	@Override
	public Class<String> getReturnType() {
		return String.class;
	}
	
	@Override
	public String toString(final @Nullable Event e, final boolean debug) {
		return "the permissions" + (getExpr().isDefault() ? "" : " of " + getExpr().toString(e, debug));
	}
}
```
But like Bensku said it can conflict with other plugins. SkQuery has this if you would like to use it. I really don't think permissions access like this should be within Skript.
